File: app/src/main/kotlin/com/absinthe/libchecker/utils/manifest/ApplicationReader.java
package com.absinthe.libchecker.utils.manifest;

import androidx.collection.ArrayMap;


import java.io.File;
import java.io.IOException;
import java.util.Map;

import pxb.android.Res_value;
import pxb.android.axml.NodeVisitor;


public class ApplicationReader {
  private final ArrayMap<String, Object> properties = new ArrayMap<>();

  private ApplicationReader(File apk) throws IOException {
    ManifestReader.acceptManifest(apk, () -> new ManifestTagVisitor(null));
  }

  public static Map<String, Object> getManifestProperties(File apk) throws IOException {
    return new ApplicationReader(apk).properties;
  }

  private ApplicationReader(byte[] bytes) throws IOException {
    ManifestReader.acceptManifest(bytes, () -> new ManifestTagVisitor(null));
  }

  public static Map<String, Object> getManifestProperties(byte[] bytes) throws IOException {
    return new ApplicationReader(bytes).properties;
  }

  private class ManifestTagVisitor extends NodeVisitor {
    public String name = null;
    public Object value = null;

    public ManifestTagVisitor(NodeVisitor child) {
      super(child);
    }

    @Override
    public NodeVisitor child(String ns, String name) {
      NodeVisitor child = super.child(ns, name);
      if ("application".equals(name)) {
        return new ApplicationTagVisitor(child);
      }
      return child;
    }

    @Override
    public void end() {
      if (name != null && value != null) {
        properties.put(name, value);
      }
      super.end();
    }

    private class ApplicationTagVisitor extends NodeVisitor {
      public String name = null;
      public Object value = null;

      public ApplicationTagVisitor(NodeVisitor child) {
        super(child);
      }

      @Override
      public void attr(String ns, String name, int resourceId, String raw, Res_value value) {
        this.name = name;
        if (value.type == Res_value.TYPE_REFERENCE) {
          this.value = value.data;
        } else {
          this.value = value.toString();
        }

        if (name != null && value.type != Res_value.TYPE_NULL) {
          properties.put(name, this.value);
        }
        super.attr(ns, name, resourceId, raw, value);
      }

      @Override
      public void end() {
        if (name != null && value != null) {
          properties.put(name, value);
        }
        super.end();
      }
    }
  }
}

File: app/src/main/kotlin/com/absinthe/libchecker/utils/manifest/ApplicationReader.kt
package com.absinthe.libchecker.utils.manifest

import androidx.collection.ArrayMap
import com.absinthe.libchecker.compat.IZipFile
import com.absinthe.libchecker.compat.ZipFileCompat
import pxb.android.Res_value
import pxb.android.axml.AxmlReader
import pxb.android.axml.AxmlVisitor
import pxb.android.axml.NodeVisitor
import timber.log.Timber
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException
import java.io.InputStream

class ApplicationReader private constructor(apk: File) {
    private val properties = ArrayMap<String, Any>()

    init {
        try {
            ZipFileCompat(apk).use { zip ->
                val inputStream = zip.getInputStream(zip.getEntry("AndroidManifest.xml"))
                val bytes = getBytesFromInputStream(inputStream)
                val reader = AxmlReader(bytes ?: ByteArray(0))
                reader.accept(object : AxmlVisitor() {
                    override fun child(ns: String?, name: String?): NodeVisitor {
                        val child = super.child(ns, name)
                        return ManifestTagVisitor(child)
                    }
                })
            }
        } catch (e: Exception) {
            Timber.e(e)
        }
    }

    companion object {
        @JvmStatic
        @Throws(IOException::class)
        fun getManifestProperties(apk: File): Map<String, Any> {
            return ApplicationReader(apk).properties
        }

        @JvmStatic
        fun getBytesFromInputStream(inputStream: InputStream?): ByteArray? {
            if (inputStream == null) return null
            
            return try {
                ByteArrayOutputStream().use { bos ->
                    val b = ByteArray(1024)
                    var n: Int
                    while (inputStream.read(b).also { n = it } != -1) {
                        bos.write(b, 0, n)
                    }
                    bos.toByteArray()
                }
            } catch (e: Exception) {
                Timber.w(e)
                null
            }
        }
    }

    private inner class ManifestTagVisitor(child: NodeVisitor?) : NodeVisitor(child) {
        var name: String? = null
        var value: Any? = null

        override fun child(ns: String?, name: String?): NodeVisitor {
            val child = super.child(ns, name)
            return if ("application" == name) {
                ApplicationTagVisitor(child)
            } else child
        }

        override fun end() {
            val tagName = name
            val tagValue = value
            if (tagName != null && tagValue != null) {
                properties[tagName] = tagValue
            }
            super.end()
        }

        private inner class ApplicationTagVisitor(child: NodeVisitor?) : NodeVisitor(child) {
            var name: String? = null
            var value: Any? = null

            override fun attr(ns: String?, name: String?, resourceId: Int, raw: String?, value: Res_value?) {
                this.name = name
                this.value = if (value?.type == Res_value.TYPE_REFERENCE) {
                    value.data
                } else {
                    value?.toString()
                }

                if (name != null && value?.type != Res_value.TYPE_NULL) {
                    this.value?.let { properties[name] = it }
                }
                super.attr(ns, name, resourceId, raw, value)
            }

            override fun end() {
                val attrName = name
                val attrValue = value
                if (attrName != null && attrValue != null) {
                    properties[attrName] = attrValue
                }
                super.end()
            }
        }
    }
}